File: lambda/audio.py
import json
import boto3
import os
import base64
import logging
from datetime import datetime

logger = logging.getLogger()
logger.setLevel(logging.INFO)

s3 = boto3.client('s3')
BUCKET = os.environ['TRAINING_BUCKET']

def get_ws_client(event):
    endpoint = f"https://{event['requestContext']['domainName']}/{event['requestContext']['stage']}"
    return boto3.client('apigatewaymanagementapi', endpoint_url=endpoint)

def send_ws_response(api_client, connection_id, message):
    try:
        api_client.post_to_connection(
            ConnectionId=connection_id,
            Data=json.dumps(message)
        )
    except Exception as e:
        logger.error(f"Failed to send WebSocket response: {str(e)}")

def handler(event, context):
    logger.info(f"Event received: {json.dumps(event)}")
    
    try:
        if 'requestContext' not in event or 'connectionId' not in event['requestContext']:
            logger.error("Invalid WebSocket event format")
            return {'statusCode': 400}
            
        connection_id = event['requestContext']['connectionId']
        route = event['requestContext']['routeKey']
        api_client = get_ws_client(event)
        
        if route == '$connect':
            return {'statusCode': 200}
            
        elif route == '$disconnect':
            return {'statusCode': 200}
            
        elif route == '$default':
            try:
                body = json.loads(event.get('body', '{}'))
            except json.JSONDecodeError:
                error_message = {'status': 'error', 'message': 'Invalid JSON in request body'}
                send_ws_response(api_client, connection_id, error_message)
                return {'statusCode': 400}
                
            if 'audio' not in body:
                error_message = {'status': 'error', 'message': 'Missing audio data'}
                send_ws_response(api_client, connection_id, error_message)
                return {'statusCode': 400}
                
            try:
                # Save audio
                audio_data = base64.b64decode(body['audio'])
                # Simple validation - check if we got actual data
                if len(audio_data) == 0:
                    error_message = {'status': 'error', 'message': 'Empty audio data'}
                    send_ws_response(api_client, connection_id, error_message)
                    return {'statusCode': 400}
                    
                timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
                # Sanitize the filename to prevent path traversal
                safe_connection_id = connection_id.replace('/', '_').replace('..', '_')
                key = f"audio/{safe_connection_id}/{timestamp}.wav"
                
                s3.put_object(
                    Bucket=BUCKET,
                    Key=key,
                    Body=audio_data
                )
                
                # Send response back through WebSocket
                response_data = {'status': 'success', 'file': key}
                send_ws_response(api_client, connection_id, response_data)
                
            except Exception as audio_error:
                logger.error(f"Error processing audio: {str(audio_error)}")
                error_message = {'status': 'error', 'message': f'Failed to process audio: {str(audio_error)}'}
                send_ws_response(api_client, connection_id, error_message)
                return {'statusCode': 500}
                
            return {'statusCode': 200}
            
    except Exception as e:
        logger.error(f"Error: {str(e)}")
        return {'statusCode': 500, 'body': json.dumps({'error': str(e)})}

File: lambda-ts/src/audio.ts
import { Context } from 'aws-lambda';
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
import { ApiGatewayManagementApiClient, PostToConnectionCommand } from '@aws-sdk/client-apigatewaymanagementapi';
import { WebSocketEvent, WebSocketResult } from './types';

// Initialize S3 client
const s3Client = new S3Client({});

/**
 * Create an API Gateway Management API client for the WebSocket
 */
function getWsClient(event: WebSocketEvent): ApiGatewayManagementApiClient {
    const endpoint = `https://${event.requestContext.domainName}/${event.requestContext.stage}`;
    return new ApiGatewayManagementApiClient({ endpoint });
}

/**
 * Send a response through the WebSocket connection
 */
async function sendWsResponse(
    apiClient: ApiGatewayManagementApiClient,
    connectionId: string,
    message: Record<string, any>
): Promise<void> {
    try {
        const command = new PostToConnectionCommand({
            ConnectionId: connectionId,
            Data: Buffer.from(JSON.stringify(message)),
        });

        await apiClient.send(command);
    } catch (error) {
        console.error('Failed to send WebSocket response:', error);
    }
}

/**
 * Handler for WebSocket requests
 */
export const handler = async (event: WebSocketEvent, context: Context): Promise<WebSocketResult> => {
    console.log('Event received:', JSON.stringify(event, null, 2));

    try {
        // Validate request context
        if (!event.requestContext || !event.requestContext.connectionId) {
            console.error('Invalid WebSocket event format');
            return { statusCode: 400 };
        }

        const connectionId = event.requestContext.connectionId;
        const route = event.requestContext.routeKey;
        const apiClient = getWsClient(event);

        // Handle connection
        if (route === '$connect') {
            return { statusCode: 200 };
        }

        // Handle disconnection
        if (route === '$disconnect') {
            return { statusCode: 200 };
        }

        // Handle default route (messages)
        if (route === '$default') {
            let body: any;

            try {
                body = event.body ? JSON.parse(event.body) : {};
            } catch (error) {
                const errorMessage = { status: 'error', message: 'Invalid JSON in request body' };
                await sendWsResponse(apiClient, connectionId, errorMessage);
                return { statusCode: 400 };
            }

            // Validate audio data
            if (!body.audio) {
                const errorMessage = { status: 'error', message: 'Missing audio data' };
                await sendWsResponse(apiClient, connectionId, errorMessage);
                return { statusCode: 400 };
            }

            try {
                // Decode base64 audio data
                const audioData = Buffer.from(body.audio, 'base64');

                // Simple validation - check if we got actual data
                if (audioData.length === 0) {
                    const errorMessage = { status: 'error', message: 'Empty audio data' };
                    await sendWsResponse(apiClient, connectionId, errorMessage);
                    return { statusCode: 400 };
                }

                // Create timestamp for the filename
                const timestamp = new Date().toISOString()
                    .replace(/[-:]/g, '')
                    .replace('T', '_')
                    .replace(/\..+/, '');

                // Sanitize the connection ID to prevent path traversal
                const safeConnectionId = connectionId.replace(/[\/\\.]/g, '_');
                const key = `audio/${safeConnectionId}/${timestamp}.wav`;

                // Get bucket name from environment variable
                const bucketName = process.env.TRAINING_BUCKET;
                if (!bucketName) {
                    throw new Error('TRAINING_BUCKET environment variable not set');
                }

                // Save audio to S3
                await s3Client.send(new PutObjectCommand({
                    Bucket: bucketName,
                    Key: key,
                    Body: audioData
                }));

                // Send success response
                const responseData = { status: 'success', file: key };
                await sendWsResponse(apiClient, connectionId, responseData);

                return { statusCode: 200 };
            } catch (audioError) {
                console.error('Error processing audio:', audioError);
                const errorMessage = {
                    status: 'error',
                    message: `Failed to process audio: ${audioError instanceof Error ? audioError.message : 'Unknown error'}`
                };
                await sendWsResponse(apiClient, connectionId, errorMessage);
                return { statusCode: 500 };
            }
        }

        // Handle unknown route
        return { statusCode: 400, body: JSON.stringify({ error: `Unknown route: ${route}` }) };
    } catch (error) {
        console.error('Error:', error);
        return {
            statusCode: 500,
            body: JSON.stringify({ error: error instanceof Error ? error.message : 'Unknown error' })
        };
    }
}; 